### PR TITLE
Performance improvements, use bulk_write during copy/delete phase

### PR DESCRIPTION
**Pull request by jamesandres**

### CODE REVIEW

File: shardmonster/sharder.py [hunk 1]
```diff
@@ -49,32 +62,34 @@ def _do_copy(collection_name, shard_key, manager):
...
+            tum_ti_tum(manager.insert_throttle)
+            manager.inc_inserted(by=result.bulk_api_result['nUpserted'])
     finally:
         cursor.close()
-    raise_last_mongo_error(new_collection.database)
 
 
-def upsert(collection, record):
-    collection.update({'_id': record['_id']}, record, w=0, upsert=True)
+def batch_of_update_ones(batch):
```
Comment (lastcraft) [inline shardmonster/sharder.py]: No idea what the name means

File: shardmonster/sharder.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import sys
 import threading
 import time
+import traceback
```
Comment (lastcraft) [inline shardmonster/sharder.py]: I assume debugging code

**Comment by lastcraft** [comment]
Sensible so far

File: shardmonster/sharder.py (hunk 2 above)
Comment (jamesandres) [reply shardmonster/sharder.py]: Nope, it's the first commit. I'm improving the exception handler.

File: shardmonster/sharder.py (hunk 1 above)
Comment (jamesandres) [reply shardmonster/sharder.py]: It's creating a list of `UpdateOne` objects. It assumes the reader knows the pymongo api. Could do `batch_of_update_one_operations`?

Comment (lastcraft) [reply shardmonster/sharder.py]: I'd say if it's the only game in town, then specifying the underlying implementation is not useful. Just go with intent until there is differentiation.

File: shardmonster/sharder.py [hunk 3]
```diff
@@ -264,21 +282,24 @@ def run(self):
...
-            delete_throttle=None, insert_throttle=None):
+            delete_throttle, insert_throttle,
+            delete_batch_size, insert_batch_size):
         self.collection_name = collection_name
         self.shard_key = shard_key
         self.new_location = new_location
         self.delete_throttle = delete_throttle
         self.insert_throttle = insert_throttle
+        self.delete_batch_size = delete_batch_size
+        self.insert_batch_size = insert_batch_size
```
Comment (lastcraft) [inline shardmonster/sharder.py]: Should these be synchronised properties? Failing that, you should check that a batch size of zero doesn't hang.

**Comment by lastcraft** [comment]
Are you going to pin pymongo to 3.1 (same as conversocial app)

File: shardmonster/sharder.py (hunk 3 above)
Comment (lastcraft) [reply shardmonster/sharder.py]: Agreed to skip give the short shelf life

**Pull request merged**